File: bfs_maze_class.py

class Tile():
    def __init__(self, x, y, z):
        self.position = [x, y, z]
        self.color = "white"
        self.adj_list = []
        self.path = []
        self.visited = False

    def __init__(self):
        self.adj_list = []
        self.path = []

    def new_adj(self, tile):
        self.adj_list.append(tile)

class Graph:
    def __init__(self):
        self.initial_tile = Tile(0, 0, 0)
        self.initial_tile.color = "silver"
        self.all_tiles = []
        self.seen_tile = {
            [0,0,0]: True
        }
    
    def new_tile(self, tile):
        self.all_tiles.append(tile)

    def clear_path(self):
        for v in self.all_tiles:
            v.path.clear()

    def bfs(self, start, end):
        Graph.clear_path()
        list = []
        list.append(start)
        visited = {}
        seen = {}
        seen[start] = True
        while not list.empty():
            node = list.pop(0)
            if not visited[node]:
                visited[node] = True
                if node == end:
                    return node.path
            for i in range(len(node.adj_list)):
                if not visited[node.adj_list[i]]:
                    list.append(node.adj_list[i])
                    seen[node.adj_list[i]] = True
                    if not node.adj_list[i]: 
                        node.adj_list[i].path.append(node)

def sharp_dist(port):
    dist = 1 #retornar distancia do sensor 
    return dist

def wall_follower():
    if (sharp_dist("right_front") > 10 and sharp_dist("right_back") > 10):
        

def check_around(m, curr_til):
    if (sharp_dist("front_left") > 10 and sharp_dist("front_right") > 10):
        next_til = Tile(curr_til.position[0] + 1, curr_til.position[1], curr_til.position[2])
        if (not m.seen[next_til.position]) and (not next_til.visited):
            m.seen[next_til.position] = True

    if (sharp_dist("back_left") > 10 and sharp_dist("back_right") > 10):
        next_til = Tile(curr_til.position[0] - 1, curr_til.position[1], curr_til.position[2])
        if (not m.seen[next_til.position]) and (not next_til.visited):
            m.seen[next_til.position] = True

    if (sharp_dist("right_front") > 10 and sharp_dist("right_back") > 10):
        next_til = Tile(curr_til.position[0], curr_til.position[1] + 1, curr_til.position[2])
        if (not m.seen[next_til.position]) and (not next_til.visited):
            m.seen[next_til.position] = True

    if (sharp_dist("left_front") > 10 and sharp_dist("left_back") > 10):
        next_til = Tile(curr_til.position[0], curr_til.position[1] - 1, curr_til.position[2])
        if (not m.seen[next_til.position]) and (not next_til.visited):
            m.seen[next_til.position] = True

if __name__ == "__main__":
    maze = Graph()
    current_tile = maze.initial_tile
    check_around(maze, current_tile)

File: bfs_maze_class.cpp
#include<bits/stdc++.h>
#include<iostream>
using namespace std;

struct Tile {
    int id_Tile;
    int position[3]; //vector(x,y,z)
	int color;
	vector<Tile> adj_list;
    vector<Tile> path;
}; 

class Graph {
	public:
		int n_nodes, n_conections;
		vector<Tile> all_tiles;
        vector<Tile> seen;

		Graph() {
			n_nodes = 1;
			n_conections = 0; 
            Tile initial; // x = 0, y = 0, z = 0
            for (int i = 0; i < 3; i++) initial.position[i] = 0;
            initial.id_Tile = 1;
		}
		
		void add_conections(Tile* s, Tile* f) { //s-start f-final
			s->adj_list.push_back(*f);
			n_conections++;
		}

        void clear_path() {
            for (int i = 0; i < n_nodes; i++) {
                all_tiles[i].path.clear();
            }
        }
		
		vector<Tile> bfs(Tile start, Tile end) { 
            clear_path();
			vector<Tile> list;
			list.push_back(start);
			map<int,bool> seen_;
			while(!list.empty()) {
				Tile node = list[0];
				list.erase(list.begin());
				if(!seen_[node.id_Tile]) {
					seen_[node.id_Tile] = true;
					if (node.id_Tile == end.id_Tile) return node.path;
				}
				for(int i = 0; i < node.adj_list.size(); i++) {
					if(!seen_[node.adj_list[i].id_Tile]) {
						list.push_back(node.adj_list[i]);
                        node.adj_list[i].path.push_back(node);
					}
				}
			}
		}
};

int sharp_dist(string port);
void check_around(Graph* m);

// x:positive to the right, y:positive to the front, z:positive up

int main() {
    Graph* maze;
    check_around(maze);
}


void check_around(Graph* m) {
    if (sharp_dist("front_left") > 10 && sharp_dist("front_right") > 10) {
        
    }
}